-1

**Failed Tests**:   Build  ClangBuild 
**Size**: This PR adds an extra 36KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7dc9f9/42135/summary.html
**COMMIT**: 86b908328648f93fb1c5a6e35642e6929d1b58f3
**CMSSW**: CMSSW_14_2_X_2024-10-11-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/46358/42135/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>--- Registered EDM Plugin: CondCoreCondDBXMLAuthenticationService
Entering library rule at src/CondCore/CondDB/plugins
>> Compiling  src/CondCore/CondDB/plugins/CondDBPyBind11Wrappers.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH='CMSSW_14_2_X_2024-10-11-1100' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_14_2_X_2024-10-11-1100' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_2_X_2024-10-11-1100/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/cms/coral/CORAL_2_3_21-24138eb286a8c6982010231211d8d80b/include/LCG -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/pcre/8.43-e34796d17981e9b6d174328c69446455/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/boost/1.80.0-477823d53efabc5118f199265eb7ab49/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/curl/7.79.0-abae79fca02c8e83ab0f428f2a91d96b/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/python3/3.9.14-c10287ae9cadff55334e60003302c349/include/python3.9 -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/lcg/root/6.30.09-c59d6b036f1cbd6988c172ba319259f1/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/tbb/v2021.9.0-e785b749a0b6cb9c66dc1d78066210e0/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/zlib/1.2.11-1a082fc322b0051b504cc023f21df178/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/fmt/10.2.1-deba157dd76fed17ac8b42adc60ff53f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/py3-pybind11/2.12.0-857784db2d3fea20ed3006c707e589a0/lib/python3.9/site-packages/pybind11/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/tinyxml2/6.2.0-68d511689c5dbd9a0d296bbb795fa052/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v2 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/CondCore/CondDB/plugins/CondDBPyBind11Interface/CondDBPyBind11Wrappers.cc.d src/CondCore/CondDB/plugins/CondDBPyBind11Wrappers.cc -o tmp/el8_amd64_gcc12/src/CondCore/CondDB/plugins/CondDBPyBind11Interface/CondDBPyBind11Wrappers.cc.o
src/CondCore/CondDB/plugins/CondDBPyBind11Wrappers.cc: In function 'std::tuple<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > cond::getDbCredentials(const std::string&, int, const std::string&)':
src/CondCore/CondDB/plugins/CondDBPyBind11Wrappers.cc:14:7: error: 'assert' was not declared in this scope
   14 |       assert(auth_path_env);
      |       ^~~~~~
src/CondCore/CondDB/plugins/CondDBPyBind11Wrappers.cc:5:1: note: 'assert' is defined in header '<cassert>'; did you forget to '#include <cassert>'?
    4 | #include <cstdlib>
  +++ |+#include <cassert>
</pre>

## Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Entering Package CondCore/CondDB
>> Entering Package CondFormats/JetMETObjects
>> Entering Package CondTools/DT
>> Entering Package CondTools/SiPixel
>> Compile sequence completed for CMSSW CMSSW_14_2_X_2024-10-11-1100
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
+ eval scram build outputlog '&&' '(python3' /data/cmsbld/jenkins/workspace/ib-run-pr-tests/cms-bot/buildLogAnalyzer.py --logDir /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_2_X_2024-10-11-1100/tmp/el8_amd64_gcc12/cache/log/src '||' 'true)'
++ scram build outputlog
>> Entering Package CondCore/CondDB
Entering library rule at src/CondCore/CondDB/plugins
>> Compiling edm plugin src/CondCore/CondDB/plugins/RelationalAuthenticationService.cc
</pre>

--- a/CondCore/CondDB/plugins/CondDBPyBind11Wrappers.cc
+++ b/CondCore/CondDB/plugins/CondDBPyBind11Wrappers.cc
@@ -2,6 +2,7 @@
 #include "CondCore/CondDB/interface/Auth.h"
 //
 #include <cstdlib>
+#include <cassert>
 
 namespace cond {
 
@@ -10,7 +11,9 @@ namespace cond {
                                                                      const std::string& authPath) {
     std::string ap = authPath;
     if (ap.empty()) {
-      ap = std::string(std::getenv(cond::auth::COND_AUTH_PATH));
+      char* auth_path_env = std::getenv(cond::auth::COND_AUTH_PATH);
+      assert(auth_path_env);
+      ap = std::string(auth_path_env);
     }
     auto ret = std::make_tuple(std::string(""), std::string(""), std::string(""));
     if (!ap.empty()) {

--- a/CondCore/CondDB/src/CoralMsgReporter.cc
+++ b/CondCore/CondDB/src/CoralMsgReporter.cc
@@ -26,6 +26,7 @@ cond::persistency::CoralMsgReporter::CoralMsgReporter()
   // Use a non-default message level?
   if (getenv("CORAL_MSGLEVEL")) {
     // Check only the first char of the environment variable
+    [[clang::suppress]]
     switch (*getenv("CORAL_MSGLEVEL")) {
       case '0':
       case 'n':

--- a/CondFormats/JetMETObjects/bin/JetResolution_t.cc
+++ b/CondFormats/JetMETObjects/bin/JetResolution_t.cc
@@ -61,8 +61,9 @@ int main(int argc, char** argv) {
   cout << "nevts:    " << nevts << endl;
   cout << "gaussian: " << doGaussian << endl << endl;
 
-  string cmssw_base(std::getenv("CMSSW_BASE"));
-  string cmssw_release_base(std::getenv("CMSSW_RELEASE_BASE"));
+  // CMSSW_BASE and CMSSW_RELEASE_BASE are always set
+  [[clang::suppress]] string cmssw_base(std::getenv("CMSSW_BASE"));
+  [[clang::suppress]] string cmssw_release_base(std::getenv("CMSSW_RELEASE_BASE"));
   string path = cmssw_base + "/src/CondFormats/JetMETObjects/data";
   struct stat st;
   if (stat(path.c_str(), &st) != 0)

--- a/CondFormats/JetMETObjects/bin/JetResolution_t.cc
+++ b/CondFormats/JetMETObjects/bin/JetResolution_t.cc
@@ -7,6 +7,7 @@
 ////////////////////////////////////////////////////////////////////////////////
 
 #include "CondFormats/JetMETObjects/interface/JetResolution.h"
+#include "FWCore/ParameterSet/interface/FileInPath.h"
 
 #include <TROOT.h>
 #include <TApplication.h>
@@ -61,21 +62,12 @@ int main(int argc, char** argv) {
   cout << "nevts:    " << nevts << endl;
   cout << "gaussian: " << doGaussian << endl << endl;
 
-  // CMSSW_BASE and CMSSW_RELEASE_BASE are always set
-  [[clang::suppress]] string cmssw_base(std::getenv("CMSSW_BASE"));
-  [[clang::suppress]] string cmssw_release_base(std::getenv("CMSSW_RELEASE_BASE"));
-  string path = cmssw_base + "/src/CondFormats/JetMETObjects/data";
-  struct stat st;
-  if (stat(path.c_str(), &st) != 0)
-    path = cmssw_release_base + "/src/CondFormats/JetMETObjects/data";
-  if (stat(path.c_str(), &st) != 0) {
-    cerr << "ERROR: tried to set path but failed, abort." << endl;
-    return 0;
-  }
-
-  string ptFileName = path + "/" + era + "_PtResolution_" + alg + ".txt";
-  string etaFileName = path + "/" + era + "_EtaResolution_" + alg + ".txt";
-  string phiFileName = path + "/" + era + "_PhiResolution_" + alg + ".txt";
+  string ptFileName =
+      edm::FileInPath("CondFormats/JetMETObjects/data/" + era + "_PtResolution_" + alg + ".txt").fullPath();
+  string etaFileName =
+      edm::FileInPath("CondFormats/JetMETObjects/data/" + era + "_EtaResolution_" + alg + ".txt").fullPath();
+  string phiFileName =
+      edm::FileInPath("CondFormats/JetMETObjects/data/" + era + "_PhiResolution_" + alg + ".txt").fullPath();
 
   cout << ptFileName << endl;
   cout << etaFileName << endl;

--- a/CondTools/DT/src/DTPerformanceHandler.cc
+++ b/CondTools/DT/src/DTPerformanceHandler.cc
@@ -78,6 +78,11 @@ void DTPerformanceHandler::getNewObjects() {
   std::map<std::string, popcon::PayloadIOV>::iterator itag =
     mp.find( dataTag );
 */
+  cond::Time_t snc = runNumber;
+  if (runNumber <= last) {
+    std::cout << "More recent data already present - skipped" << std::endl;
+    return;
+  }
 
   DTPerformance* dtPerf = new DTPerformance(dataTag);
 
@@ -125,15 +130,7 @@ void DTPerformanceHandler::getNewObjects() {
 */
 
   //for each payload provide IOV information (say in this case we use since)
-  cond::Time_t snc = runNumber;
-  if (runNumber > last)
-    m_to_transfer.push_back(std::make_pair(dtPerf, snc));
-  else {
-    delete dtPerf;
-    std::cout << "More recent data already present - skipped" << std::endl;
-  }
-
-  return;
+  m_to_transfer.push_back(std::make_pair(dtPerf, snc));
 }
 
 std::string DTPerformanceHandler::id() const { return dataTag; }

--- a/CondTools/DT/src/DTPerformanceHandler.cc
+++ b/CondTools/DT/src/DTPerformanceHandler.cc
@@ -128,8 +128,10 @@ void DTPerformanceHandler::getNewObjects() {
   cond::Time_t snc = runNumber;
   if (runNumber > last)
     m_to_transfer.push_back(std::make_pair(dtPerf, snc));
-  else
+  else {
+    delete dtPerf;
     std::cout << "More recent data already present - skipped" << std::endl;
+  }
 
   return;
 }

--- a/CondTools/DT/src/DTPerformanceHandler.cc
+++ b/CondTools/DT/src/DTPerformanceHandler.cc
@@ -78,7 +78,6 @@ void DTPerformanceHandler::getNewObjects() {
   std::map<std::string, popcon::PayloadIOV>::iterator itag =
     mp.find( dataTag );
 */
-  cond::Time_t snc = runNumber;
   if (runNumber <= last) {
     std::cout << "More recent data already present - skipped" << std::endl;
     return;
@@ -130,7 +129,7 @@ void DTPerformanceHandler::getNewObjects() {
 */
 
   //for each payload provide IOV information (say in this case we use since)
-  m_to_transfer.push_back(std::make_pair(dtPerf, snc));
+  m_to_transfer.emplace_back(dtPerf, runNumber);
 }
 
 std::string DTPerformanceHandler::id() const { return dataTag; }

--- a/CondTools/DT/src/DTStatusFlagHandler.cc
+++ b/CondTools/DT/src/DTStatusFlagHandler.cc
@@ -76,8 +76,6 @@ void DTStatusFlagHandler::getNewObjects() {
   std::map<std::string, popcon::PayloadIOV>::iterator itag =
     mp.find( dataTag );
 */
-  cond::Time_t snc = runNumber;
-
   if (runNumber <= last) {
     std::cout << "More recent data already present - skipped" << std::endl;
     return;
@@ -128,7 +126,7 @@ void DTStatusFlagHandler::getNewObjects() {
 */
 
   //for each payload provide IOV information (say in this case we use since)
-  m_to_transfer.push_back(std::make_pair(stFlag, snc));
+  m_to_transfer.emplace_back(stFlag, runNumber);
 }
 
 std::string DTStatusFlagHandler::id() const { return dataTag; }

--- a/CondTools/DT/src/DTStatusFlagHandler.cc
+++ b/CondTools/DT/src/DTStatusFlagHandler.cc
@@ -76,6 +76,12 @@ void DTStatusFlagHandler::getNewObjects() {
   std::map<std::string, popcon::PayloadIOV>::iterator itag =
     mp.find( dataTag );
 */
+  cond::Time_t snc = runNumber;
+
+  if (runNumber <= last) {
+    std::cout << "More recent data already present - skipped" << std::endl;
+    return;
+  }
 
   DTStatusFlag* stFlag = new DTStatusFlag(dataTag);
 
@@ -122,15 +128,7 @@ void DTStatusFlagHandler::getNewObjects() {
 */
 
   //for each payload provide IOV information (say in this case we use since)
-  cond::Time_t snc = runNumber;
-  if (runNumber > last)
-    m_to_transfer.push_back(std::make_pair(stFlag, snc));
-  else {
-    delete stFlag;
-    std::cout << "More recent data already present - skipped" << std::endl;
-  }
-
-  return;
+  m_to_transfer.push_back(std::make_pair(stFlag, snc));
 }
 
 std::string DTStatusFlagHandler::id() const { return dataTag; }

--- a/CondTools/DT/src/DTStatusFlagHandler.cc
+++ b/CondTools/DT/src/DTStatusFlagHandler.cc
@@ -125,8 +125,10 @@ void DTStatusFlagHandler::getNewObjects() {
   cond::Time_t snc = runNumber;
   if (runNumber > last)
     m_to_transfer.push_back(std::make_pair(stFlag, snc));
-  else
+  else {
+    delete stFlag;
     std::cout << "More recent data already present - skipped" << std::endl;
+  }
 
   return;
 }

--- a/CondTools/SiPixel/plugins/SiPixelDynamicInefficiencyReader.cc
+++ b/CondTools/SiPixel/plugins/SiPixelDynamicInefficiencyReader.cc
@@ -270,11 +270,13 @@ void SiPixelDynamicInefficiencyReader::analyze(const edm::Event& e, const edm::E
       mismatch++;
     }
     for (unsigned int i = 0; i < pu_det; i++) {
+      [[clang::suppress]]
       if (_pu_scale[i] != 0 && _pu_scale_conf[i] != 0 && _pu_scale[i] == _pu_scale_conf[i]) {
         //printf("Config match! detid %x\t db_pu_scale %f\tconf_pu_scale %f\n",detid.rawId(),_pu_scale[i],_pu_scale_conf[i]);
         pu_match++;
         break;
       }
+      [[clang::suppress]]
       if (_pu_scale[i] != 0 && _pu_scale_conf[i] != 0 && _pu_scale[i] != _pu_scale_conf[i]) {
         //printf("Config mismatch! detid %x\t db_pu_scale %f\tconf_pu_scale %f\n",detid.rawId(),_pu_scale[i],_pu_scale_conf[i]);
         pu_mismatch++;

--- a/CondTools/SiPixel/plugins/SiPixelDynamicInefficiencyReader.cc
+++ b/CondTools/SiPixel/plugins/SiPixelDynamicInefficiencyReader.cc
@@ -197,8 +197,8 @@ void SiPixelDynamicInefficiencyReader::analyze(const edm::Event& e, const edm::E
     }
 
     //DB PU factor calculation
-    unsigned int pu_iterator = 0;
-    for (it_pu = map_pufactor.begin(); it_pu != map_pufactor.end(); it_pu++, pu_iterator++) {
+    it_pu = map_pufactor.begin();
+    for (unsigned int pu_iterator = 0; pu_iterator < pu_det; it_pu++, pu_iterator++) {
       const DetId mapid = DetId(it_pu->first);
       if (mapid.subdetId() != detid.subdetId())
         continue;
@@ -270,13 +270,11 @@ void SiPixelDynamicInefficiencyReader::analyze(const edm::Event& e, const edm::E
       mismatch++;
     }
     for (unsigned int i = 0; i < pu_det; i++) {
-      [[clang::suppress]]
       if (_pu_scale[i] != 0 && _pu_scale_conf[i] != 0 && _pu_scale[i] == _pu_scale_conf[i]) {
         //printf("Config match! detid %x\t db_pu_scale %f\tconf_pu_scale %f\n",detid.rawId(),_pu_scale[i],_pu_scale_conf[i]);
         pu_match++;
         break;
       }
-      [[clang::suppress]]
       if (_pu_scale[i] != 0 && _pu_scale_conf[i] != 0 && _pu_scale[i] != _pu_scale_conf[i]) {
         //printf("Config mismatch! detid %x\t db_pu_scale %f\tconf_pu_scale %f\n",detid.rawId(),_pu_scale[i],_pu_scale_conf[i]);
         pu_mismatch++;

--- a/CondTools/SiPixel/plugins/SiPixelGainCalibrationRejectNoisyAndDead.cc
+++ b/CondTools/SiPixel/plugins/SiPixelGainCalibrationRejectNoisyAndDead.cc
@@ -225,7 +225,7 @@ void SiPixelGainCalibrationRejectNoisyAndDead::fillDatabase(const edm::EventSetu
     float gainforthiscol[2];
     float pedforthiscol[2];
     int nusedrows[2];
-    int nrowsrocsplit;
+    int nrowsrocsplit = 0;
     if (record_ == "SiPixelGainCalibrationOfflineRcd")
       nrowsrocsplit = theGainCalibrationDbInputOffline_->getNumberOfRowsToAverageOver();
     if (record_ == "SiPixelGainCalibrationForHLTRcd")

